modules/resolve.js

```diff
@@ -0,0 +1,83 @@
+export const resolveLocation = (location, base) => {
+  if (typeof location === 'string') {
+    return resolve(location, base)
+  } else {
+    location.pathname = resolve(location.pathname, base)
```

Comment:
> It does not work when to includes a hash or search string.

`resolve-pathname` is only designed to resolve the pathname portion of the URL. Leaving out search and hash was intentional. As you say, stripping and re-adding that stuff after the pathname is correctly resolved is the easy part :) I think I'd rather do that in the router codebase somewhere and just let `resolve-pathname` do the one thing it's good at.

As for non-compliance with the RFC, I agree with you. Having a `<Route path="c">` nested under a `<Route path="/a/b">` should match `/a/b/c`. However, I think we should still be able to use `resolve-pathname` for that by just adding a trailing slash to the parent route's path before passing it in, e.g. `resolvePathname(parentMatch.path + '/', path)`.